tests/core/actions/utils.py

```diff
@@ -0,0 +1,16 @@
+import json
+
+
+def validate_action_str(action_str, action_parameters):
+    """
+    Parse the action string as JSON within an exec call and verify that it matches the original action parameters.
+    Note that we need the exec call to mimic the same behavior in the rules container.
+
+    If this triggers a syntax error, the exec call is set up incorrectly and needs to be fixed.
+    If this triggers a JSON decode error, the action string is badly formatted.
+    If this triggers an assertion error, the deserialized action JSON does not match the original action parameters.
+    """
+    try:
+        exec(f'import json; assert json.loads("{action_str}") == {action_parameters}')
+    except (SyntaxError, json.JSONDecodeError, AssertionError) as e:
+        raise Exception(f"Error occurred during action string validation.")
```

Comment:
more details about exception, you should bubble up e.

Reply:
e doesn't provide a useful message, it just lists the exception (since the code is triggered in another process)

Reply:
hence why I updated the docstring about what each possible exception means.

Reply:
do we need to do 'exec'? 
Its all pythonic statement, it can be verified in same process, No?

Reply:
See the docstring. The point is to mimic the same behavior in the rules container, where the JSON is parsed within an exec call.

The reason why this matters is because two backslashes are needed for strings if this is executed within an exec call. By simply parsing json in this process, this triggers json validation error.

Reply:
Makes sense.